controllers/kubeadmconfig_controller.go

```diff
@@ -263,6 +267,83 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	return ctrl.Result{}, nil
 }
 
+// reconcileDiscovery ensure that config.JoinConfiguration.Discovery is properly set for the joining node.
+// The implementation func respect user provided discovery configurations, but in case some of them are missing, a valid BootstrapToken object
+// is automatically injected into config.JoinConfiguration.Discovery.
+// This allows to simplify configuration UX, by providing the option to delegate to CABPK the configuration of kubeadm join discovery.
+func (r *KubeadmConfigReconciler) reconcileDiscovery(cluster *capiv1alpha2.Cluster, config *cabpkv1alpha2.KubeadmConfig) error {
+	log := r.Log.WithValues("kubeadmconfig", fmt.Sprintf("%s/%s", config.Namespace, config.Name))
+
+	// if config already contains a file discovery configuration, respect it without further validations
+	if config.Spec.JoinConfiguration.Discovery.File != nil {
+		return nil
+	}
+
+	// otherwise it is necessary to ensure token discovery is properly configured
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
+	}
+
+	// if BootstrapToken already contains an APIServerEndpoint, respect it; otherwise inject the APIServerEndpoint endpoint defined in cluster status
+	//TODO(fp) might be we want to validate user provided APIServerEndpoint and warn/error if it doesn't match the api endpoint defined at cluster level
+	apiServerEndpoint := config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint
+	if apiServerEndpoint == "" {
+		if len(cluster.Status.APIEndpoints) == 0 {
+			return errors.New("Control plane already exists for the cluster but cluster.Status.APIEndpoints are not set")
+		}
+
+		// NB. CABPK only uses the first APIServerEndpoint defined in cluster status if there are multiple defined.
+		apiServerEndpoint = fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port)
```

Comment:
i know the controlPlaneEndpoint should *not* have `https://` prefixed. Is that different for the bootstrapToken apiserverendpoint?

Reply:
aligned to the same format of controlPlane endpoint (AFAIK bootstrapToken apiserverendpoint accepts both)